### PR TITLE
Implement PatchMutation::ApplyToLocalView

### PR DESCRIPTION
**Pull request by rsgowman**

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
Suggestion: Do you want to merge this to held_acks_master and port the new version of the code directly from Android?

**Comment by rsgowman** [comment]
Answering Sebastian's question (well; summarizing from an offline discussion): This PR is pretty minimal, and missing quite a lot. It's just got the stuff required to proceed with local serialization, and otherwise has large holes in the implementation. So we're going to have to take a second pass at this later to fill the rest in, at which point, (I hope) all the held write acks stuff should be done.

That said, it's largely ported from android (rather than ios) so is mostly following Sebastian's suggestion already.

File: Firestore/core/test/firebase/firestore/testutil/testutil.h [hunk 1]
```diff
@@ -43,6 +44,12 @@ namespace firebase {
 namespace firestore {
 namespace testutil {
 
+/**
+ * A string sentinel that can be used with PatchMutation() to mark a field for
+ * deletion.
+ */
+constexpr const char* kDeleteSentinel = "<DELETE>";
```
Comment (zxu123) [inline Firestore/core/test/firebase/firestore/testutil/testutil.h]: Isn't this already defined somewhere not specific to test? Or actually you want to test the sentinel is the same?

File: Firestore/core/test/firebase/firestore/testutil/testutil.h [hunk 2]
```diff
@@ -131,6 +138,39 @@ inline std::unique_ptr<model::SetMutation> SetMutation(
       model::Precondition::None());
 }
 
+inline std::unique_ptr<model::PatchMutation> PatchMutation(
```
Comment (zxu123) [inline Firestore/core/test/firebase/firestore/testutil/testutil.h]: The logic is not that simple and I assume it is used in a few places. How about remove inline and save compilation time?

File: Firestore/core/src/firebase/firestore/model/mutation.cc [hunk 3]
```diff
@@ -67,6 +68,57 @@ std::shared_ptr<const MaybeDocument> SetMutation::ApplyToLocalView(
...
+}
+
+std::shared_ptr<const MaybeDocument> PatchMutation::ApplyToLocalView(
+    const std::shared_ptr<const MaybeDocument>& maybe_doc,
+    const MaybeDocument*,
+    const Timestamp&) const {
+  VerifyKeyMatches(maybe_doc.get());
+
+  if (!precondition().IsValidFor(maybe_doc.get())) {
+    if (maybe_doc)
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/model/mutation.cc]: Done.

File: Firestore/core/test/firebase/firestore/testutil/testutil.h (hunk 1 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/testutil/testutil.h]: No, it only seems to appear in the testutil class (at least for android, objc, and now c++)

File: Firestore/core/test/firebase/firestore/testutil/testutil.h (hunk 2 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/testutil/testutil.h]: Done.

**Pull request merged**